### ISSUE TITLE
Support for QLoRA

### ISSUE THREAD

**Issue by zcsctc** [oobabooga/text-generation-webui#2352]
**Description**
The QLoRA needs a  bitsandbytes library >=0.39.0. 
![image](https://github.com/oobabooga/text-generation-webui/assets/56668272/61c87fce-3aed-4dc3-b898-bd290a7db709)
However, the project can't run when the library is >=0.38.1 according to this [issuse](https://github.com/TimDettmers/bitsandbytes/issues/175)
![image](https://github.com/oobabooga/text-generation-webui/assets/56668272/f0ad2fb2-31e8-47de-bdbe-bd7abc9c7593)

**Comment by jepjoo** [comment]
bitsandbytes==0.39.0 in requirements.txt already

**Comment by ReDXeoL** [comment]
yesterday everything worked fine for me, but today I downloaded the latest version of the Windows installer and it seems that bitsandbytes 0.39.0 was updated and now it is not compatible with my nvida rtx 3090 gpu despite downloading everything wanted by cuda gpu
![Captura](https://github.com/oobabooga/text-generation-webui/assets/127172146/c3281d5b-ae55-480c-93c7-d074331b0245)

**Comment by NickWithBotronics** [comment]
Has anyone been able to get qlora running and training through the web ui?

**Comment by GamingDaveUk** [comment]
How do you switch to qlora? I have yet to get any lora training to work with the gptq 4bit models i use.

**Comment by fpena06** [comment]
> Has anyone been able to get qlora running and training through the web ui?

I'm wondering the same thing. I hope someone can shed some light on this. I started [this](https://github.com/oobabooga/text-generation-webui/discussions/2722) discussion.

**Comment by NickWithBotronics** [comment]
Yeah I do it all the time! Use load in 4 bit with transformers, use nf4 bf16. Then in the training page change the optimizer to Adam8bit. The rest of the parameters configure you have to test and learn what works best. For your card and if you are using a cut off length of of 512 on a 24 gig card you can do a batch size of like 25 with a rank of up to 256! If you do this it won't make the model only preform 512 outputs it will still work with 2048 tokens unless you were to do a full fine tune (not Lora) it really won't loose this ability during training.
